### DISCUSSION TITLE
Can we run cpptools.exe as a daemon so it can index and parse my C++ projet?

### DISCUSSION THREAD

**Discussion by stephane-archer** [microsoft/vscode-cpptools#8127] [General]
Then Visual Code can be start and the work is already done. (I have a really large codebase)

**Comment by bobbrow** [comment]
Not currently. `cpptools.exe` does more than just project indexing. That functionality would have to be separated out in order to work and I we're not currently investigating doing that.  We would also have to get the legal team involved since our license currently restricts the usage of our binaries to within the context of VS Code.

**Reply by stephane-archer** [reply]
From my understanding, `cpptools.exe` is a `Language Servers` which communicates over a protocol to other processes. What would be the technical problem to do a daemon that "indexing" project? 
When I mean "indexing" I mean to do some work to accelerate requests to `cpptools.exe` if it's doing much more things, I don't mind, it has all night to make my request fast when I arrive in the morning at work. 
Regarding the legal team, it's about using `cpptools.exe` with Vscode. I'm maybe wrong but currently, every time I close Vscode I lose the "indexing work" done by `cpptools.exe`?
Maybe "indexing" is not the appropriate word, do you have a better word for what I tried to describe?

**Reply by bobbrow** [reply]
You shouldn't be losing the indexing work when you close VS Code.  It should be saved in a database that is reused the next time you open the same workspace.  If you are seeing the indexer delete the database or start over every time, please open a bug so we can start an investigation.

**Comment by stephane-archer** [comment]
I'm not sure if it's starting over or just continuing where it left off. Is
there any way to know? Can I query/ask the server? Where is located the
database? Can I query/ask the database?

**Reply by sean-mcmanus** [reply]
You can set the databaseFilename, see https://code.visualstudio.com/docs/cpp/c-cpp-properties-schema-reference#_browse-properties . If you set C_Cpp.loggingLevel to Debug it'll show a message if the database is getting reset.

The database can get reset if the cpptools process exits in an abnormal way within 5 seconds of a database write, i.e. we treat that as a potentially corrupt database.

The LSP is a protocol that is only supported to communicate with VS Code, and "indexing" isn't part of the LSP protocol.

**Reply by stephane-archer** [reply]
I will do some testing in the next coming days and let you know.

I'm not sure I understand correctly the last phrase.
"GoToDefinition" is part of the LSP protocol. To be able to answer this question quickly, you need to "index"/'analyze" the code in advance right?
Things can be hidden in macros and multiple classes can have functions with the same name.
Can can't just "grep" when you receive a request "GoToDefinition" right?

Maybe "indexing" is not the appropriate word, any suggestion?

**Reply by sean-mcmanus** [reply]
Yes, we need to index and parse the code in advance for go to definition to work. You might be able to use the LSP if you send the correct initialize and textDocument/openFile etc. messages to cpptools. Go to Definition isn't needed to build most of the database since most files will be parsed from the root "in the background".
